src/3d/qgsrulebased3drenderer.cpp

```diff
@@ -0,0 +1,470 @@
+
+  // call recursively
+  Q_FOREACH ( Rule *rule, mChildren )
+  {
+    // Don't process else rules yet
+    if ( !rule->isElse() )
+    {
+      RegisterResult res = rule->registerFeature( feature, context, handlers );
+      // consider inactive items as "registered" so the else rule will ignore them
+      willRegisterSomething |= ( res == Registered || res == Inactive );
+      registered |= willRegisterSomething;
+    }
+  }
+
+  // If none of the rules passed then we jump into the else rules and process them.
+  if ( !willRegisterSomething )
+  {
+    Q_FOREACH ( Rule *rule, mElseRules )
+    {
+      registered |= rule->registerFeature( feature, context, handlers ) != Filtered;
+    }
+  }
+
+  if ( !mIsActive )
+    return Inactive;
+  else if ( registered )
+    return Registered;
+  else
+    return Filtered;
+}
+
+
+bool QgsRuleBased3DRenderer::Rule::isFilterOK( QgsFeature &f, Qgs3DRenderContext &context ) const
+{
+  if ( ! mFilter || mElseRule )
+    return true;
+
+  context.expressionContext().setFeature( f );
+  QVariant res = mFilter->evaluate( &context.expressionContext() );
+  return res.toInt() != 0;
+}
+
+
+////////////////////
+
+
+QgsRuleBased3DRenderer::QgsRuleBased3DRenderer( QgsRuleBased3DRenderer::Rule *root )
+  : mRootRule( root )
+{
+}
+
+QgsRuleBased3DRenderer::~QgsRuleBased3DRenderer()
+{
+  delete mRootRule;
+}
+
+QgsRuleBased3DRenderer *QgsRuleBased3DRenderer::clone() const
+{
+  Rule *rootRule = mRootRule->clone();
+
+  // normally with clone() the individual rules get new keys (UUID), but here we want to keep
+  // the tree of rules intact, so that other components that may use the rule keys work nicely (e.g. map themes)
+  rootRule->setRuleKey( mRootRule->ruleKey() );
+  RuleList origDescendants = mRootRule->descendants();
+  RuleList clonedDescendants = rootRule->descendants();
+  Q_ASSERT( origDescendants.count() == clonedDescendants.count() );
+  for ( int i = 0; i < origDescendants.count(); ++i )
+    clonedDescendants[i]->setRuleKey( origDescendants[i]->ruleKey() );
+
+  QgsRuleBased3DRenderer *r = new QgsRuleBased3DRenderer( rootRule );
+  r->mLayerRef = mLayerRef;
+  return r;
+}
+
+void QgsRuleBased3DRenderer::setLayer( QgsVectorLayer *layer )
+{
+  mLayerRef = QgsMapLayerRef( layer );
+}
+
+QgsVectorLayer *QgsRuleBased3DRenderer::layer() const
+{
+  return qobject_cast<QgsVectorLayer *>( mLayerRef.layer );
+}
+
+Qt3DCore::QEntity *QgsRuleBased3DRenderer::createEntity( const Qgs3DMapSettings &map ) const
+{
+  QgsVectorLayer *vl = layer();
+
+  if ( !vl )
+    return nullptr;
+
+  Qgs3DRenderContext context( map );
+  // TODO: add expr. context
+
+  RuleToHandlerMap handlers;
+  mRootRule->createHandlers( vl, handlers );
+
+  QSet<QString> attributeNames;
+  mRootRule->prepare( context, attributeNames, handlers );
+
+  QgsFeatureRequest req;
+  req.setDestinationCrs( map.crs(), map.transformContext() );
+  req.setSubsetOfAttributes( attributeNames, vl->fields() );
```

Comment:
I noticed you haven't ported the equivalent of QgsRuleBasedRenderer::filter for setting an expression based filter on the request based on the visible rules -- this is a huge performance boost for rule based renderer, and would probably help here too (and is likely just more copy+past from rule based renderer)

Reply:
I have based the implementation on rule-based labeling which does not have it... yeah I can add support for this but would leave it up for a later time if you don't mind (preparation time here is typically dominated by tessellation)